File: src/python/runner.py
import os.path
import subprocess
import json
import sys

def real_path(file):
	return os.path.join(sys.path[0], file)

def generate_grid(default_item, size):
	return [[dict(default_item) for _ in range(size[1])] for _ in range(size[0])]

def assertm(bool_, message):
	if !bool_:
		raise Exception(message)

def verify_type(data, type_):
	assertm(type(data) == type_, 'Invalid data format.')

def verify_in_grid(coords):
	assertm(
		(
			coords[0] >= 0 and
			coords[0] < GRID_SIZE[0] and
			coords[1] >= 0 and
			coords[1] < GRID_SIZE[1]
		),
		'Coords "{}" not in grid'.format(coords)
	)

def verify_key(dict_, key):
	assertm(
		key in dict_,
		'Key "{}" not found'.format(key)
	)

class Bot:
	def __init__(self, language, file):
		self.language = language
		self.file = file
		if self.language == 'python':
			self.command = 'python3'
		else:
			raise Exception('Invalid language')

	def run(self, data):
		return subprocess.run(
			[
				self.command,
				real_path(self.file),
				'<',
				json.dumps(data).encode('utf-8')
			],
			stdin=subprocess.PIPE,
			stdout=subprocess.PIPE
		).stdout.decode('utf-8')

GRID_SIZE = [10, 10]
grid = generate_grid(
	{
		'type': 'plain',
	},
	GRID_SIZE,
)
bots = [
	Bot('python', 'sample-bot.py'),
	Bot('python', 'sample-bot.py'),
]
unit_commands = {
	'soldier': ['move']
}

template = [
	((0, 0), False): {
		('type', True): '',

	}
]

def verify_type(obj, template):
	assertm(type(obj) == type(template), 'Invalid data value type.')
	if type(obj) == dict:
		for key, required in template:
			if required:
				assertm(key in obj, 'Missing data key "{}"'.format(key))
		for key in obj:
			assertm(type(key) == type(list(template.keys())[0])), 'Invalid data key type.')
			verify_type(obj[key], template[list(template.keys())[0]])
	elif type(obj) == list or tuple:
		for val in obj:
			verify_type(val, template[0])

for i in range(10):
	for bot in bots:
		commands = bot.run(grid)
		verify_type(commands, template)

		for coords in commands:
			command = commands[coords]
			verify_in_grid(coords)
			unit = grid[coords[0]][coords[1]]

			if command['type'] not in unit_commands[unit['type']]:
				raise Exception('No command "{}" for unit "{}"'.format(
					command['type'], unit['type']
				))

			if command['type'] == 'move':

File: src/node-new/runner.js
const child_process = require('child_process')
const vld = require('./vld.js')
const util = require('util')

function generateGrid(default_item, size) {
  const grid = []
  for (let x = 0; x < size[0]; x++) {
    grid[x] = []
    for (let y = 0; y < size[1]; y++) {
      grid[x][y] = Object.assign({}, default_item)
    }
  }
  return grid
}

function inGrid(coords) {
  return (
    coords[0] >= 0 &&
    coords[0] < GRID_SIZE[0] &&
    coords[1] >= 0 &&
    coords[1] < GRID_SIZE[1]
  )
}

function toCoords(string) {
  return string.split(',').map(char => parseInt(char))
}

class Bot {
  constructor(language, file) {
    this.language = language
    this.file = file
    if (this.language == 'python') {
      this.command = 'python3'
    } else {
      throw new Error('Invalid language')
    }
  }

  run(data) {
    return child_process.run(
      `${self.command} ${real_path(this.file)} < ${JSON.stringify(data)}`
    )
  }
}


GRID_SIZE = [100, 100]
UNIT_COMMANDS = {
  'soldier': ['move']
}

grid = generateGrid(
  {
    type: 'plain',
  },
  GRID_SIZE,
)
grid[10][10] = {
  unit: 'soldier',
}
// console.log(grid)
bots = [
  new Bot('python', 'sample-bot.py'),
  new Bot('python', 'sample-bot.py'),
]


const validator = new vld.Validator({
  type: 'object',
  key: {
    type: 'string',
    regex: /^\d*,\d*$/g,
    check: coords => inGrid(toCoords(coords))
  },
  value: [
    vld.equal({
      type: ['move', 'attack'],
      direction: ['left', 'right', 'up', 'down'],
    }),
  ],
  check: commands => {
    let success = true
    for (let key in commands) {
      const command = commands[key]
      const coords = toCoords(key)
      command.unit = grid[coords[0]][coords[1]].unit
      success = success &&
        command.unit in UNIT_COMMANDS &&
        UNIT_COMMANDS[command.unit].includes(command.type)
    }
    return success
  }
})

console.log(util.inspect(validator.template, false, null, true))

console.log(validator.validate({
  '10,10': 1
}))


// for (let i = 0; i < 10; i++) {
// 	for (let bot of bots) {
// 		commands = bot.run(grid)

//     for (let id of commands) {

//     }
//   }
// }
